feat: update dependencies, install angular-material and convert the a…

jzolnowski opened the pull request:
- update dependencies
- install angular-material and convert the application to be similar to angularjs-angular-hybrid-angular-cli 
- improve webpack configuration
- migrate to eslint

NOTE: the [e2e PR](https://github.com/xlts-dev/angularjs-webpack-babel/pull/19) should be merged first

jzolnowski committed:
feat: update dependencies, install angular-material and convert the application to be similar to angularjs-angular-hybrid-angular-cli, improve webpack configuration, provide production build, fix lint errors

gkalpak commented on src/app/version-stamp.component.js:
```diff
@@ -0,0 +1,42 @@
+  this.angularJsIcon = require('../assets/angularjs.svg');
+  this.angularjsMaterialIcon = require('../assets/angularjs-material.png');
+  this.versions = {
+    angularjs: angular.version,
```
OOC, should we do `$window.angular` (for consistency)?

jzolnowski replied:
Changed

gkalpak commented on src/app/app-angularjs.component.js:
```diff
@@ -0,0 +1,8 @@
+             <footer>
+                <app-version-stamp></app-version-stamp>
+             </footer>`,
+  controller: class AppAngularjsCtrl {},
```
Could we omit `controller` altogether?

jzolnowski replied:
Removed

gkalpak commented on package.json:
```diff
@@ -38,11 +38,12 @@
+    "start:prod-local": "cd dist && http-server -p 4200 -c-1 -b",
     "prettier": "prettier --write \"**/*.{js,json,css,scss,less,md,ts,html}\"",
     "lint": "eslint .",
     "update": "npx npm-check -u",
```
Since we now have `npm-check` as a direct dependency, could this be just `"update": "npm-check -u"`?

jzolnowski replied:
Changed

gkalpak commented on config/plugins.mjs:
```diff
@@ -0,0 +1,31 @@
+import HtmlWebpackPlugin from 'html-webpack-plugin';
+import MiniCssExtractPlugin from 'mini-css-extract-plugin';
+import { paths, isProd } from './helpers.mjs';
+import {CleanWebpackPlugin} from 'clean-webpack-plugin';
+import webpack from 'webpack';
```
Let's order imports alphabetically (by import path):

```suggestion
import {CleanWebpackPlugin} from 'clean-webpack-plugin';
import HtmlWebpackPlugin from 'html-webpack-plugin';
import MiniCssExtractPlugin from 'mini-css-extract-plugin';
import webpack from 'webpack';
import { paths, isProd } from './helpers.mjs';
```

jzolnowski replied:
Reordered

gkalpak commented on config/plugins.mjs:
```diff
@@ -0,0 +1,31 @@
+const commonPlugins = [
+  new CleanWebpackPlugin(),
+  new HtmlWebpackPlugin({
+    title: 'XLTS for AngularJS with Webpack 5',
```
Let's make this more future proof by removing the version:

```suggestion
    title: 'XLTS for AngularJS with Webpack',
```

jzolnowski replied:
Version removed

gkalpak commented on config/common.mjs:
```diff
@@ -0,0 +1,10 @@
+import { rules } from './rules.mjs';
+import { plugins } from './plugins.mjs';
```
Let's order imports alphabetically (by import path):

```suggestion
import { plugins } from './plugins.mjs';
import { rules } from './rules.mjs';
```

jzolnowski replied:
Reordered

gkalpak commented on config/common.mjs:
```diff
@@ -0,0 +1,10 @@
+import { rules } from './rules.mjs';
+import { plugins } from './plugins.mjs';
+
+export const common = {
```
I think it would be better to have a more descriptive name (both for the exported const and for the file itself).
Maybe `commonConfig` or `commonWebpackConfig` instead of just `common`.

This would give more context on the file's/code's purpose, without having to find the usages.

jzolnowski replied:
Changed to `webapck-common-config` to be consistent with `webapck-dev` and `webpack-prod` configs

gkalpak commented on package.json:
```diff
@@ -56,45 +57,55 @@
-    "core-js": "^3.20.3"
+    "angular-animate": "npm:@xlts.dev/angular-animate@1.9.1",
+    "angular-aria": "npm:@xlts.dev/angular-aria@1.9.1",
+    "angular-material": "npm:@xlts.dev/angular-material@^1.2.4",
```
Why are we introducing AngularJS Material?
I thought we decided that we didn't want to introduce any new dependencies (including AngularJS Material).

jzolnowski replied:
I know that such a comment was added to the tracking document. However, the changes I made were done before this document was written. I wanted the repository consistent with` angularjs-angular-hybrid-angular-cli`, but it only applied to `AngularJS `without `TypeScript`. The added code improved the project configuration, for example, adding `parserOptions `in the `eslint `configuration.

gkalpak replied:
I do have a preference for keeping things simple with these demo apps (i.e. not include extra dependencies unless they are necessary or we want to demonstrate something), but I don't feel too strongly about it.

I'll defer to @Splaktar, because we had discussed that we don't want to introduce a new dependency.
If this is fine with him, it's fine with me.

Splaktar replied:
I agree with @gkalpak, and I specifically spoke with @jzolnowski and let him know that we should not add AngularJS Material to this repo.

That said since the work is already done, I will look at it and decide if we should keep it. I will need to run the app locally to do that, so I won't be able to take care of that right now.

Splaktar replied:
Based on the lack of functionality in the original version of this app and the fact that this work is already done, I think it is acceptable to add this additional dependency.

gkalpak commented on src/index.html:
```diff
@@ -1,17 +1,24 @@
-      <test-directive></test-directive>
-      <test-component></test-component>
-    </section>
+    <meta name="description" content="XLTS for AngularJS with Webpack 5 app demo." />
+    <meta name="keywords" content="XLTS for AngularJS, AngularJS Material, Webpack 5" />
```
Let's remove version numbers to make this more future proof.

jzolnowski replied:
Version removed

Splaktar replied:
This still says "Webpack 5".

gkalpak commented on src/index.html:
```diff
@@ -1,17 +1,24 @@
 <!DOCTYPE html>
-<html ng-app="app" lang="en">
+<html lang="en-US" ng-csp>
```
OOC, why introduce `ng-csp`? It feels a bit outside the scope of this repo.
(If we decided to go ahead with this change, we should manually import `angular-csp.css`.)

jzolnowski replied:
It's a good practice to not use `eval()`, `Function(string)`, and similar functions to dynamically create and execute code from strings as well as inline resources, such as inline `<script>` and `<style>` elements.

It's true that to use the `ngHide` or `ngCloak `directive, you need to import` angular-csp.css` manually, but in this project, we don't use it, so I don't know if it makes sense to import this file out of the box.

gkalpak replied:
You are right that it is a good practice in general, but I feel it might be outside the scope of this repo (whose purpose is to demonstrate how to set app a basic app).

I don't feel strongly about it, so leaving it up to you.
I do think, though, that if we keep `ngCsp` we should include `angular-csp.css` for a couple of reasons:
- It is a tiny file, so does not tangibly affect the loading of the page.
- It is not only about `ngCloak` and `ngHide`/`ngShow`. It also has styles for some kinds of animations (`.ng-anchor`, `.ng-animate-shim`).
- Even if we don't use one of the affected features directly, it is possible that one of our dependencies does.  
  (For example, the developer should not care to know if AngularJS Material relies on animations that require `.ng-anchor`/`.ng-animate-shim`.)
- Even if we don't rely on any of the features right now, it is easy to start relying in the future (either directly or indirectly via a new dependency).

Splaktar replied:
It looks like `ng-csp` was removed. Resolving this.

gkalpak commented on gulpfile.ts:
```diff
@@ -0,0 +1,17 @@
+const gulp = require('gulp');
```
How is `gulp` invoked in this repo? Is it run as part of `npm build`/`npm start` or should it be run manually?

jzolnowski replied:
Add a new script `"build:templates": "gulp templates"`. Now when running `start` or `build` script, it lunch first` npm run build:templates`

jzolnowski replied:
<img width="706" alt="image" src="https://user-images.githubusercontent.com/5073455/220364070-ef5ffb9f-163a-44ed-a096-9877f6655871.png">

gkalpak commented on .eslintrc.json:
```diff
@@ -1,26 +1,236 @@
   "globals": {
-    "TEST_MODE": true
+    "angular": true,
+    "go": true
```
What is the `go` global?

jzolnowski replied:
Removed

gkalpak commented on src/app/templates/templates.js:
```diff
@@ -0,0 +1,10 @@
+// eslint-disable-next-line
```
Should this file be checked in?
If we run the `templates` gulp task as part of the build (which we should imo), then could we git-ignore this file?

jzolnowski replied:
Added to `.gitignore`

gkalpak commented on src/app/templates/index.js:
```diff
@@ -0,0 +1,3 @@
+import templates from './templates';
```
The `templates.js` file does not have any exports. What are we importing here?

jzolnowski replied:
Good catch! I removed the entire templates folder and import cached templates directly in `app-angularjs.module.js`

gkalpak commented on src/app/version-stamp.component.js:
```diff
@@ -0,0 +1,42 @@
+
+    <md-divider id="version-stamp-divider"></md-divider>
+    <div class="version-container">
+      <span class="version"><img src="{{ $ctrl.angularJsIcon }}" alt="" height="24px"> AngularJS v{{ $ctrl.versions.angularjs.full }} ({{ $ctrl.versions.angularjs.codeName }})</span>
+      <span class="version"><img src="{{ $ctrl.angularjsMaterialIcon }}" alt="" height="24px"> AngularJS Material v{{ $ctrl.versions.md }}</span>
```
It would be better to style images via CSS, because the hot module replacement seems to result in having some generic `img` styles (from AngularJS Material? - not sure) to override the `height` attribute with a `height: auto` CSS rule (which in turn results in the AngularJS image getting pretty big).

jzolnowski replied:
Good to know:) thank for the info! Moved image styles to CSS

gkalpak reviewed (changes requested):
For future PRs:
This kind of PR is usually much easier to review if it is broken up into multiple commits with one type of change per commit (for example, one commit with all changes related to updating dependencies, one commit related to improving the webpack config, one related to improving the ESLint config, etc.).

jzolnowski committed:
fix: apply changes after George's review

jzolnowski committed:
fix(lint): fix unexpected token lint issue

jzolnowski commented:
@gkalpak PR ready for another round of review

gkalpak reviewed (changes requested):
Just a couple of comments.
Otherwise lgtm 👍

Great stuff :100:

gkalpak commented on package.json:
```diff
@@ -36,16 +36,24 @@
-    "start": "webpack-dev-server --open --mode development",
+    "pre-commit": "lint-staged && npm run lint",
+    "build:templates": "gulp templates",
+    "start": "npm run build:templates && webpack-dev-server --open --mode development --config config/webpack.dev.mjs",
```
I just realized that this command will not pick up template changes while it is running. I suggest we use something like [npm-run-all][1] to run `gulp watch` and `webpack-dev-server` in parallel.

[1]: https://www.npmjs.com/package/npm-run-all

jzolnowski committed:
Remove ngCsp directive , add gulp watch script

jzolnowski committed:
Add version stamp icons heigh attribute

gkalpak commented on src/app/version-stamp.component.js:
```diff
@@ -0,0 +1,43 @@
+
+    <md-divider id="version-stamp-divider"></md-divider>
+    <div class="version-container">
+      <span class="version"><img src="{{ $ctrl.angularJsIcon }}" alt="" height="24"> AngularJS v{{ $ctrl.versions.angularjs.full }} ({{ $ctrl.versions.angularjs.codeName }})</span>
+      <span class="version"><img src="{{ $ctrl.angularjsMaterialIcon }}" alt="" height="24"> AngularJS Material v{{ $ctrl.versions.md }}</span>
```
We are setting the height in CSS, so we don't need it here:

```suggestion
      <span class="version"><img src="{{ $ctrl.angularJsIcon }}" alt=""> AngularJS v{{ $ctrl.versions.angularjs.full }} ({{ $ctrl.versions.angularjs.codeName }})</span>
      <span class="version"><img src="{{ $ctrl.angularjsMaterialIcon }}" alt=""> AngularJS Material v{{ $ctrl.versions.md }}</span>
```

gkalpak commented on package.json:
```diff
@@ -36,16 +36,26 @@
+    "pre-commit": "lint-staged && npm run lint",
+    "build:templates": "gulp templates",
+    "start:server": "webpack-dev-server --open --mode development --config config/webpack.dev.mjs",
+    "gulp:watch": "gulp watch",
```
Nit: I would name this something more descriptive, such as `watch:templates`.

jzolnowski commented:
@gkalpak PR ready for another round of review

gkalpak commented on package.json:
```diff
@@ -36,16 +36,26 @@
+    "build:templates": "gulp templates",
+    "start:server": "webpack-dev-server --open --mode development --config config/webpack.dev.mjs",
+    "gulp:watch": "gulp watch",
+    "start": "npm run build:templates && npm-run-all -p start:server gulp:watch",
```
Nit: This can be shortened as:

```suggestion
    "start": "npm run build:templates && run-p start:server gulp:watch",
```

gkalpak reviewed (changes requested):
Almost ready :tada:

jzolnowski committed:
Remove no needed height attribute in the HTML element, rename gulp:watch to watch:templates

jzolnowski commented:
@gkalpak PR ready for final review 😄

gkalpak reviewed (approved):
LGTM (pending a decision on [this discussion][1] - I've pinged Michael to take a look at it) 🚀

[1]: https://github.com/xlts-dev/angularjs-webpack-babel/pull/18/files#r1124280976

Splaktar reviewed (changes requested):
- [x] I get the following errors when running `gulp templates`:

```
> gulp templates

[22:31:11] Failed to load external module ts-node/register
[22:31:11] Error: Cannot find module 'ts-node/register' from '/Users/splaktar/Git/xlts/angularjs-webpack-babel'
[22:31:11] Failed to load external module typescript-node/register
[22:31:11] Error: Cannot find module 'typescript-node/register' from '/Users/splaktar/Git/xlts/angularjs-webpack-babel'
[22:31:11] Failed to load external module typescript-register
[22:31:11] Error: Cannot find module 'typescript-register' from '/Users/splaktar/Git/xlts/angularjs-webpack-babel'
[22:31:11] Failed to load external module typescript-require
[22:31:11] Error: Cannot find module 'typescript-require' from '/Users/splaktar/Git/xlts/angularjs-webpack-babel'
[22:31:11] Failed to load external module sucrase/register/ts
[22:31:11] Error: Cannot find module 'sucrase/register/ts' from '/Users/splaktar/Git/xlts/angularjs-webpack-babel'
[22:31:11] Failed to load external module @babel/register
[22:31:11] Error: Cannot find module '@babel/register' from '/Users/splaktar/Git/xlts/angularjs-webpack-babel'
[22:31:12] Using gulpfile ~/Git/xlts/angularjs-webpack-babel/gulpfile.ts
[22:31:12] Starting 'templates'...
[22:31:12] Finished 'templates' after 17 ms
```
Is this expected?

- [x] The footer text and images aren't properly aligned. This isn't a problem in the angularjs-angular-material-hybrid demo.

![Screenshot 2023-03-09 at 22 36 52](https://user-images.githubusercontent.com/3506071/224217509-191c8095-b389-4dae-9533-f27f433a8b32.png)

jzolnowski committed:
fix: Failed to load external module errors

jzolnowski committed:
fix: version-stamp.component styles

jzolnowski commented:
Fixed `gulp templates` errors and fixed styles in the `versionStampComponent`

jzolnowski committed:
fix: imports order

Splaktar commented on src/styles.scss:
```diff
@@ -0,0 +1,9 @@
+@import '~angular-material/angular-material.scss';
+
+body {
+  height: 100%;
+}
+body {
+  margin: 0;
+  font-family: Roboto, 'Helvetica Neue', sans-serif;
+}
```
Why isn't this combined?

gkalpak replied:
Actually, this seems to still not be resolved.

jzolnowski committed:
feat(test): add playwright configuration and e2e tests for tabs and buttons (#19)

- `babel-eslint` with `@babel/eslint-parser`
- fix eslint issues
- replace functions by classes
- remove baseURL from playwright config
- set port to 8080
- improve package.json test scripts
- remove unused click methods in Page Object file
- add jasmine to eslint env
- add webpack as karma framework
- change task order of package.json test script
- remove `test:unit` command from the 'test' script
- improve eslint configuration

Splaktar commented:
Please rebase and resolve conflicts.

jzolnowski committed:
fix(e2e): increase timeout to 1min

jzolnowski commented:
Branch up to date with master

jzolnowski committed:
feat(test): provide unit test example for tabs component, webpack configuration for unit test purposes, fix karma config

Splaktar reviewed (dismissed):
- [x] Please respond to https://github.com/xlts-dev/angularjs-webpack-babel/pull/18#discussion_r1135800244. I know it's auto hidden by GitHub 😦

jzolnowski committed:
fix: remove version from webpack in index.html

jzolnowski commented:
> * [ ]  Please respond to [#18 (comment)](https://github.com/xlts-dev/angularjs-webpack-babel/pull/18#discussion_r1135800244). I know it's auto hidden by GitHub 😦

Sorry, I did not notice that the version repeats in several places - previously I removed it from the title. Now I have removed from the description and keywords

jzolnowski committed:
chore: set engine-strict to true, set engine npm version to 9 and node to above or equal 18

jzolnowski commented:
I also set the `engine-strict` flag in `.npmrc`. I updated the npm and node versions in the `package.json` engines: `npm v9`, `node>=18`. I tested if the package installation, start and testing of the application works correctly. Everything looks fine

gkalpak commented on config/plugins.mjs:
```diff
@@ -0,0 +1,31 @@
+      filename: '[name].css',
+      chunkFilename: '[id].css',
+    }),
+    new webpack.HotModuleReplacementPlugin(),
```
AFAICT, [since `webpack-dev-server` v4.0.0][1] (where `devServer.hot` defaults to `true`), we no longer need to include this plugin. (It is automatically included.)
Leaving it produces a warning in the terminal.

[1]: https://webpack.js.org/guides/hot-module-replacement/#enabling-hmr

gkalpak commented on config/webpack.test.js:
```diff
@@ -0,0 +1,52 @@
+const path = require('path');
```
I would add a comment here explaining why we need this `.js` file (and can't reuse other configs).

gkalpak commented on config/webpack.test.js:
```diff
@@ -0,0 +1,52 @@
+const HtmlWebPackPlugin = require('html-webpack-plugin');
+
+module.exports = () => ({
+  entry: {},
```
This produces the following warning in the terminal:

```
karma-webpack does not currently support custom entries, if this is something you need,
consider opening an issue.
ignoring attempt to set the entry option...
```

I would remove it or comment it out and add a comment explaining why there is no `entry`.

gkalpak commented on karma.config.js:
```diff
@@ -1,20 +1,18 @@
     preprocessors: {
+      'src/main.js': ['webpack'],
       'src/index.tests.js': ['webpack'],
       'src/**/*.test.js': ['webpack'],
```
This line shouldn't be needed. The `*.test.js` files are processed as part of `index.tests.js` afaict.

gkalpak commented on .npmrc:
```diff
@@ -1,2 +1,3 @@
 @xlts.dev:registry=https://registry.xlts.dev/
 //registry.xlts.dev/:always-auth=true
+engine-strict=true
```
This is a bit to strict imo.
To my experience, leaving this off is a better DX.

gkalpak commented on src/app/tabs.component.test.js:
```diff
@@ -0,0 +1,23 @@
+  );
+
+  it('should be defined', () => {
+    const ctrl = $componentController('appTabs', { $scope: scope }, { $digestCount: 0 });
```
We should not set the 3rd argument. It is used to simulate `bindToController` bindings, which there are none in this case.

This is covering up the fact that the component's controller definition is incorrect and `$digestCount` is not properly initialized in the controller as we would expect :grin:

gkalpak commented on karma.config.js:
```diff
@@ -35,9 +33,9 @@ module.exports = config => {
+    autoWatch: false, // if true, Karma captures browsers, runs the tests and exits
     browsers: ['ChromeHeadless'],
-    singleRun: false,
+    singleRun: true,
```
Nit:
It is more idiomatic to leave `autoWatch: true` and `singleRun: false` by default (or just remove them altogether, since the default is to run in watch mode), because most of the time people will want to run the unit tests in watch mode.

Then, if you want to run the tests once and exit (as we do in the `test` script), you can pass the `--single-run` option.

gkalpak commented on src/app/tabs.component.js:
```diff
@@ -0,0 +1,19 @@
+  controller: TabsCtrl,
+};
+
+class TabsCtrl {
```
This is not right. Classes are not hoisted in the same way that functions do. As a result, `controller: TabsCtrl` is the same as `controller: undefined`.

You should either move the class definition above the `tabsComponent` object initialization or define the class inline (as you do with `VersionStampCtrl`):

```suggestion
  controller: class TabsCtrl {
    ...
  },
};
```

gkalpak commented on src/app/tabs.component.test.js:
```diff
@@ -0,0 +1,23 @@
+
+  it('should be defined', () => {
+    const ctrl = $componentController('appTabs', { $scope: scope }, { $digestCount: 0 });
+    scope.$watch('ctrl.$digestCount', () => ctrl.$digestCount++);
```
This is also not right. I don't understand the purpose of it, but it does not seem to belong in this test :grin:

gkalpak commented on src/app/tabs.component.test.js:
```diff
@@ -0,0 +1,23 @@
+
+    expect(ctrl.$digestCount).toEqual(1);
+  });
+});
```
All in all, I would simplify this whole test file as follows:

```js
describe('TabsCtrl', () => {
  let $componentController;
  let $rootScope;

  beforeEach(angular.mock.module('AngularJSApp'));
  beforeEach(angular.mock.inject((_$componentController_, _$rootScope_) => {
    $componentController = _$componentController_;
    $rootScope = _$rootScope_;
  }));

  it('should be defined', () => {
    const ctrl = $componentController('appTabs');
    expect(ctrl.$digestCount).toBe(0);

    $rootScope.$digest();
    // Expect `2`, because AngularJS always runs on more digest at the end to
    // to make sure there are no new changes.
    expect(ctrl.$digestCount).toBe(2);
  });
});
```

Commit:
chore: update .nvmrc to use node v18

Commit:
fix: unit test for tabs.component.test, improve karma configuration, remove engine-strict flag, add useful comments to webpack.test.js, remove HotModuleReplacementPlugin from plugins.mjs, fix tabsComponent object initialization

jzolnowski commented:
@gkalpak Thank you for the great explanation of the causes of errors in `tabs.component.test`. 
Ready for another round 😄

gkalpak commented on karma.config.js:
```diff
@@ -1,20 +1,18 @@
     preprocessors: {
+      'src/main.js': ['webpack'],
       'src/index.tests.js': ['webpack'],
       'src/**/*.test.js': ['webpack'],
```
This line shouldn't be needed. The `*.test.js` files are processed as part of `index.tests.js` afaict.

gkalpak replied:
To clarify, I am referring to line 17 (`src/**/*.test.js`), not line 15 (`src/main.js`) :grin:

gkalpak commented on src/app/tabs.component.test.js:
```diff
@@ -0,0 +1,22 @@
+    expect(ctrl.$digestCount).toBe(0);
+
+    $rootScope.$digest();
+    // Expect `2`, because AngularJS always runs on more digest at the end to
```
```suggestion
    // Expect `2`, because AngularJS always runs one more digest at the end to
```

My bad. I had the same typo in [my comment][1] 😇

[1]: https://github.com/xlts-dev/angularjs-webpack-babel/pull/18#discussion_r1138760713

Commit:
fix(test): change comment about  cycle in the tabs.component.test

Commit:
fix(test): remove reference to main.js from karma configuration

gkalpak commented on karma.config.js:
```diff
@@ -1,17 +1,14 @@
+    files: ['src/main.js', 'src/index.tests.js'],
 
     // Preprocess test index and test files using webpack (will run babel)
     preprocessors: {
```
Don't we need to preprocess `src/main.js`?
I would expect `preprocessors` to be:

```
preprocessors: {
  'src/main.js': ['webpack'],
  'src/index.tests.js': ['webpack'],
}
```

Commit:
fix(test): add main.js refernce to karma config preprocessors, rermove src/**/*.test.js not needed refernce

gkalpak commented on package.json:
```diff
@@ -36,66 +36,96 @@
     "lint": "eslint .",
-    "update": "npx npm-check -u",
-    "prepare": "husky install"
+    "test:unit": "karma start karma.config.js",
+    "test:unit:single": "karma start karma.config.js --single-run",
```
The karma jobs require the templates to have been built. Can you make sure that (a) the templates are built before the jobs run and (b) the templates are watched and rebuilt when Karma runs in watch mode? :pray:

Commit:
fix(test): build tempaltes before the jobs run, watch tempaltes and rebuilt when Karma runs in watch mode

gkalpak commented on package.json:
```diff
@@ -36,66 +36,97 @@
     }
   ],
-  "main": "index.js",
+  "main": "main.js",
```
This doesn't look right. We should remove this property altogether, since there is no main script for this repo.

gkalpak commented on package.json:
```diff
@@ -36,66 +36,97 @@
-    "test": "karma start karma.config.js",
-    "build": "webpack --mode production",
-    "prettier": "prettier --write \"**/*.{js,json,css,scss,less,md,ts,html}\"",
+    "pre-commit": "lint-staged && npm run lint",
```
Why do we need to run both `lint-staged` and `npm run lint`?

gkalpak commented on package.json:
```diff
@@ -36,66 +36,97 @@
     "lint": "eslint .",
-    "update": "npx npm-check -u",
-    "prepare": "husky install"
+    "test:unit": "karma start karma.config.js",
+    "test:unit:watch": "npm run build:templates && run-p test:unit watch:templates",
+    "test:unit:single": "npm run build:templates && karma start karma.config.js --single-run",
```
```suggestion
    "test:unit:single": "run-s build:templates \"test:unit -- --single-run\"",
```

gkalpak replied:
Now that the names of the scripts have changed, we also need to update `README.md` accordingly.

gkalpak commented on package.json:
```diff
@@ -36,66 +36,97 @@
+    "test:unit:watch": "npm run build:templates && run-p test:unit watch:templates",
+    "test:unit:single": "npm run build:templates && karma start karma.config.js --single-run",
+    "build": "npm run build:templates && webpack --mode production --config config/webpack.prod.mjs",
+    "start:prod-local": "cd dist && http-server -p 4200 -c-1 -b",
```
Why not:
```suggestion
    "start:prod-local": "http-server dist -p 4200 -c-1 -b",
```

gkalpak reviewed (changes requested):
@jzolnowski, please also make sure all commits are signed :pray:

Commit:
fix: remove main property from package.json, update test:unit:single and start:prod-local scripts, remove lint job from package.json pre-commit script, update README script description

jzolnowski commented:
> @jzolnowski, please also make sure all commits are signed 🙏

All commits signed 🎉

jzolnowski commented:
Branch rebased with `master`

gkalpak commented on package.json:
```diff
@@ -36,66 +36,96 @@
+    "test:e2e": "playwright test",
+    "test:e2e:debug": "playwright test --project chromium --headed",
+    "test:e2e:report": "npx playwright show-report",
+    "test": "run-s build:templates lint test:unit:single test:e2e",
```
Unless I am missing something, we shouldn't need explicitly building the templates here:

```suggestion
    "test": "run-s lint test:unit:single test:e2e",
```

gkalpak commented on package.json:
```diff
@@ -36,66 +36,96 @@
     "eslint-plugin-import-order": "^2.1.4",
-    "file-loader": "^6.2.0",
-    "html-loader": "3.1.0",
+    "eslint-config-prettier": "^8.6.0",
+    "eslint-plugin-angular": "^4.1.0",
```
These two should be placed above `eslint-plugin-import-order`.

gkalpak commented on README.md:
````diff
@@ -40,8 +40,38 @@ For production
 ```
 
-For testing
+For single run unit testing
````
To be a bit more clear, I might change this to something like:

```suggestion
For running the unit tests once
```

gkalpak commented on README.md:
````diff
@@ -40,8 +40,38 @@ For production
+npm run test:unit:single
+```
+
+For watch unit testing
````
To be a bit more clear, I might change this to something like:

```suggestion
For running the unit tests in watch mode
```

gkalpak commented on README.md:
````diff
@@ -40,8 +40,38 @@ For production
+npm run test:unit:watch
+```
+
+For lint
````
```suggestion
For linting
```

gkalpak commented on README.md:
````diff
@@ -40,8 +40,38 @@ For production
+npm run test:e2e
+```
+
+For test the app (lint, unit and e2e testing)
````
```suggestion
For running all checks/tests for the app (lint, unit and e2e tests)
```

gkalpak commented on README.md:
```diff
@@ -40,8 +40,38 @@ For production
 npm build
```
Not directly related to this PR, but shouldn't this be

```suggestion
npm run build
```

gkalpak commented on README.md:
````diff
@@ -40,8 +40,38 @@ For production
+For lint
+
+```
+npm lint
````
```suggestion
npm run lint
```

gkalpak commented on README.md:
````diff
@@ -40,8 +40,38 @@ For production
+For test the app (lint, unit and e2e testing)
+
+```
+npm run test
````
This could also be:

```suggestion
npm test
```

Commit:
chore: update README script description, remove build:templates task from package.json test script, reoder devDeps in package.json

jzolnowski commented:
@gkalpak thank you for the review. Ready for another round 😄

gkalpak reviewed (approved):
LGTM :tada:

Commit:
chore: set package.json engine.node field to v16, generate new package-lock-json, set nvmrc node to v16

Commit:
refactor(styles): combine styles related to body element

jzolnowski commented:
`engines.node` field in `package.json` set to v16
